OSDOCS-12022: Dynamic plugin updates (#82215)
openshift/openshift-docs: modules/dynamic-plugin-localization.adoc

// Module included in the following assemblies:
//
// * web_console/dynamic-plugin/overview-dynamic-plugin.adoc

:_mod-docs-content-type: Procedure
[id="dynamic-plugin-api_{context}"]
= Translating messages with react-i18next

The link:https://github.com/openshift/console-plugin-template[plugin template] demonstrates how you can translate messages with link:https://www.i18next.com/[react-i18next].

.Prerequisites
* You must have the plugin template cloned locally.
* You must be running the {product-title} web console in a container. You can use either Docker or Podman 3.2.0+.
* You must have `oc` installed.

.Procedure

> . Prefix the name with `plugin__` to avoid any naming conflicts. For example, the plugin template uses the `plugin__console-plugin-template` namespace. You can use the `useTranslation` hook, for example:

Reviewer: While it does avoid conflicts, we actually require the `plugin__` prefix. The messages won't load otherwise.

I'd try make it clear that the `plugin__console-plugin-template` is only the default and must be updated when you rename your plugin. (This is mentioned below, but still a little unclear in my opinion. It might be good to give an example like `plugin__my-plugin`.)

// Module included in the following assemblies:
//
// * web_console/dynamic-plugin/overview-dynamic-plugin.adoc

:_mod-docs-content-type: PROCEDURE
[id="dynamic-plugin-api_{context}"]
= Translating messages with react-i18next

The link:https://github.com/openshift/console-plugin-template[plugin template] demonstrates how you can translate messages with link:https://www.i18next.com/[react-i18next].

.Prerequisites
* You must have the plugin template cloned locally.

> * Optional: To test your plugin locally, run the {product-title} web console in a container. You can use either Docker or Podman 3.2.0+.

Reviewer 2: I think we are supposed to use "or later" rather than `+`

.Procedure

. Prefix the name with `plugin\__` to avoid any naming conflicts. The plugin template uses the `plugin__console-plugin-template` namespace by default, and you must update when you rename your plugin for example, `plugin__my-plugin`.  You can use the `useTranslation` hook, for example:
+
[source,tsx,subs="+quotes,+macros"]
----
conster Header: React.FC = () ++=>++ {
  const ++{ t } = useTranslation('plugin__console-demo-plugin');++
  ++return <h1>{t('Hello, World!')}</h1>;++
++};++
----
+
[IMPORTANT]
====
You must match the `i18n` namespace with the name of the `ConsolePlugin` resource.
====

> . Set the `spec.i18n.loadType` field based on needed behavior. `Preload` will load all plugin's localization resources from the i18n namespace after the dynamic plugin during loading. For example, `plugin__console-demo-plugin`

Reviewer 2: - Consider defining `loadType: Preload` after the block with a callout rather than in the step text. 
- Could likewise make the "For example, plugin__console-demo-plugin" fragment into a title for the block
- Also present tense :)

```suggestion
. Set the `spec.i18n.loadType` field based on needed behavior. The `Preload` setting loads all the plugin's localization resources from the `i18n` namespace after the dynamic plugin during loading. For example, `plugin__console-demo-plugin`
```

+
[source,yaml,subs="+quotes,+macros"]
----
spec:
  backend:
    service:
      basePath: ++/++
      name: console-demo-plugin
      namespace: console-demo-plugin
      port: 9001
    type: Service
  displayName: OpenShift Console Demo Plugin
  i18n:
    loadType: Preload
----

> . Use the format `++%plugin__console-plugin-template~My Label%++` for labels in `console-extensions.json`. The console will replace the value with the message for the current language from the `plugin__console-plugin-template` namespace. For example:

Reviewer 2: ```suggestion
. Use the format `++%plugin__console-plugin-template~My Label%++` for labels in `console-extensions.json`. The console replaces the value with the message for the current language from the `plugin__console-plugin-template` namespace. For example:
```

+
[source,json,subs="+quotes,+macros"]
----
  {
    ++"type": "console.navigation/section"++,
    ++"properties": {++
      ++"id": "admin-demo-section"++,
      ++"perspective": "admin"++,
      ++"name": "%plugin__console-plugin-template~Plugin Template%"++
    }
  }
----

> . Include a comment in a TypeScript file for link:https://github.com/i18next/i18next-parser[i18next-parser] to add the message from `console-extensions.json` to your message catalog. For example:

Reviewer 2: Have you cleared this non-RH GH repo link with PX?

Author: iiuc it is part of the support agreement we have for the template.

+
[source,tsx,subs="+quotes,+macros"]
----
++// t('plugin__console-demo-plugin~Demo Plugin')++
----

> . To update the JSON files in the `locales` folder of the plugin template when adding or changing message run the following command:

Reviewer 2: ```suggestion
. To update the JSON files in the `locales` folder of the plugin template when adding or changing a message, run the following command:
```

// Module included in the following assemblies:
//
// * web_console/dynamic-plugin/overview-dynamic-plugin.adoc

:_mod-docs-content-type: PROCEDURE
[id="dynamic-plugin-api_{context}"]
= Translating messages with react-i18next

> The link:https://github.com/openshift/console-plugin-template[plugin template] demonstrates how you can translate messages with link:https://www.i18next.com/[react-i18next]. The `v1alpha1` `apiVersion` of `ConsolePlugin` supports `console.openshift.io/use-i18n` annotation.

Reviewer: @jhadvig Have we removed `v1alpha1`?

Author: Removed it since I dont think it is really necessary to add. I can add it back though if you think it is essential information to have.

.Prerequisites
* You must have the plugin template cloned locally.

> * You must be running the {product-title} web console in a container. You can use either Docker or Podman 3.2.0+.

Reviewer: This is necessary to test your plugin locally, but isn't strictly required for the steps in this section.

> * You must have `oc` installed.

Reviewer: `oc` isn't needed

// Module included in the following assemblies:
//
// * web_console/dynamic-plugin/overview-dynamic-plugin.adoc

:_mod-docs-content-type: Procedure
[id="dynamic-plugin-api_{context}"]

> = Translating messages with react-i18next

Author: Hey @spadgett I noticed I missed some things on i18n procedure. I was trying to make it more of a procedure than how it is in the README. Can you let me know if this is the correct order and if these are accurate. Thanks! 😄

The link:https://github.com/openshift/console-plugin-template[plugin template] demonstrates how you can translate messages with link:https://www.i18next.com/[react-i18next]. The `v1alpha1` `apiVersion` of `ConsolePlugin` supports `console.openshift.io/use-i18n` annotation.

.Prerequisites
* You must have the plugin template cloned locally.
* You must be running the {product-title} web console in a container. You can use either Docker or Podman 3.2.0+.
* You must have `oc` installed.

.Procedure

> . Indicate the `ConsolePlugin` contains localization resources. If the annotation is set to `true`, the localization resources from the i18n namespace named after the dynamic plugin are preloaded. If the annotation is set to any other value or is missing on the `ConsolePlugin` resource, localization resources are not preloaded.

Reviewer: This is no longer an annotation. `spec.i18n.loadType: Preload` enables preloading (although this is mentioned below so maybe we delete this paragraph?)

+

> [source,terminal]

Reviewer: ```suggestion
[source,yaml]
```

----
spec:
  backend:
    service:
      basePath: /
      name: console-demo-plugin
      namespace: console-demo-plugin
      port: 9001
    type: Service
  displayName: OpenShift Console Demo Plugin
  i18n:
    loadType: Preload
  proxy:

----

> . Prefix the name with `plugin__` to avoid any naming conflicts. The plugin template uses the `plugin__console-plugin-template` namespace by default, and you must update when you rename your plugin for example, `plugin__my-plugin`.  You can use the `useTranslation` hook , for example:

Reviewer: ```suggestion
. Prefix the name with `plugin__` to avoid any naming conflicts. The plugin template uses the `plugin__console-plugin-template` namespace by default, and you must update when you rename your plugin for example, `plugin__my-plugin`.  You can use the `useTranslation` hook, for example:
```

+
[source,tsx]
----
conster Header: React.FC = () => {
  const { t } = useTranslation('plugin__console-demo-plugin');
  return <h1>{t('Hello, World!')}</h1>;
};
----

[IMPORTANT]
====
You must match the `i18n` namespace with the name of the `ConsolePlugin` resource.
====

. Set the `spec.i18n.loadType` field based on needed behavior. `Preload` will load all plugin's localization resources from the i18n namespace after the dynamic plugin during loading. For example, `plugin__console-demo-plugin`
+
[source,yaml]
----
spec:
  backend:
    service:
      basePath: /
      name: console-demo-plugin
      namespace: console-demo-plugin
      port: 9001
    type: Service
  displayName: OpenShift Console Demo Plugin
  i18n:
    loadType: Preload

>   proxy:

Reviewer: I would leave the incomplete `proxy` out of this example
